### PR TITLE
preserve residual in transformer norm_first

### LINKED ISSUE #61165: Incorrect pre-LN in the transformer model
The recent pull request #60593 introduced `norm_first`, i.e., pre-LayerNorm, into transformer models. 

However, its implementation does NOT match the one proposed in various papers. 
Also, it differs from the one used in ESPnet. So I think its implementation is incorrect.

---

## Current implementation from #60593

https://github.com/pytorch/pytorch/blob/54673fc944110a615bf4eadb231bd4bb114ea22f/torch/nn/modules/transformer.py#L324-L332

The line
```
src = self.norm1(src)
```
is problematic as it does not save a copy of the input.

## The one used by ESPnet
See https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transformer/encoder_layer.py#L71
```python
        residual = x
        if self.normalize_before:
            x = self.norm1(x)
```

## The one from the paper "On Layer Normalization in the Transformer Architecture"

The following figure is from https://arxiv.org/pdf/2002.04745.pdf

Screen Shot 2021-07-02 at 1 28 10 PM

You can see from (b), we should save a copy of the input `x_l`

## The one from the paper "Transformers without Tears: Improving the Normalization of Self-Attention"

The equation shown below is from https://arxiv.org/pdf/1910.05895.pdf

Screen Shot 2021-07-02 at 1 30 12 PM

---

@bhosmer 
@jbschlosser

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @mruberry @bhosmer @cpuhrsch

### PR DESCRIPTION
**Pull request by bhosmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61692 preserve residual in transformer norm_first**

Fixes #61165, code refactored a bit for clarity.

Differential Revision: [D29706830](https://our.internmc.facebook.com/intern/diff/D29706830)

### CODE REVIEW

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -321,25 +321,28 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None, src_key_paddin
         Shape:
             see the docs in Transformer class.
         """
+
+        # see Fig. 1 of https://arxiv.org/pdf/2002.04745v1.pdf
+
+        # self-attention block
+        def mha(x):
+            x = self.self_attn(x, x, x, attn_mask=src_mask, key_padding_mask=src_key_padding_mask)[0]
```
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py:329]: nit: need_weights could be set to False if you're indexing at 0 for the return value

Comment (bhosmer) [reply torch/nn/modules/transformer.py:329]: Ah yeah nice catch!

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -321,25 +321,31 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None, src_key_paddin
...
+            x = self.self_attn(x, x, x,
+                               attn_mask=src_mask,
+                               key_padding_mask=src_key_padding_mask,
+                               need_weights=False)[0]
+            return self.dropout1(x)
+
+        # feed forward block
+        def ff(x):
+            x = self.linear2(self.dropout(self.activation(self.linear1(x))))
+            return self.dropout2(x)
```
Comment (jbschlosser) [inline torch/nn/modules/transformer.py:338]: Looks like these function definitions break JIT :(

Could try adding them as private functions instead (e.g. `_mha_block()`, `_ffn_block()`)

**Review by jbschlosser — COMMENTED** [review]
Thanks for the fix! Does a similar update need to be made in `TransformerDecoderLayer` as well?

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (bhosmer) [reply torch/nn/modules/transformer.py]: argh. ok made them methods, that seems more orderly than private functions. But maybe there's a reason why private functions would be better? Happy to change if so, my preference is just aesthetic.

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (jbschlosser) [reply torch/nn/modules/transformer.py]: Ah yeah, sorry I should have said "private methods". I think they're good as methods, but they should be marked as private with the leading underscore so no one depends on them if we want to refactor later on

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

**Review by jbschlosser — APPROVED** [review]
LGTM! More correct & much more readable :)

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29706830).

**Comment by navvye** [comment]
need_weights always should have been set as a parameter for the _sa_block function. This unintentionally made outputting attention_maps using TransformerEncoderLayers impossible.

**Pull request closed without merging**